libs/utils/include/utils/FixedCircularBuffer.h

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (C) 2023 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef TNT_UTILS_FIXEDCIRCULARBUFFER_H
+#define TNT_UTILS_FIXEDCIRCULARBUFFER_H
+
+#include <utils/debug.h>
+
+#include <memory>
+#include <optional>
+#include <stddef.h>
+#include <type_traits>
+
+namespace utils {
+
+template<typename T>
+class FixedCircularBuffer {
+public:
+    explicit FixedCircularBuffer(size_t capacity)
+        : mData(std::make_unique<T[]>(capacity)), mCapacity(capacity) {}
+
+    size_t size() const noexcept { return mSize; }
+    size_t capacity() const noexcept { return mCapacity; }
+    bool full() const noexcept { return mCapacity > 0 && mSize == mCapacity; }
+    bool empty() const noexcept { return mSize == 0; }
+
+    /**
+     * Push v into the buffer. If the buffer is already full, removes the oldest item and returns
+     * it. If this buffer has no capacity, simply returns v.
+     * @param v the new value to push into the buffer
+     * @return if the buffer was full, the oldest value which was displaced
+     */
+    std::optional<T> push(T v) noexcept {
+        if (mCapacity == 0) {
```

Comment:
just curious, why would you ever have a mCapacity== 0, also I think it'll segfault when you do `% mCapacity`?

Reply:
`mCapacity` is set to 0 in the case where the use-after-free detection is turned off. In that case, `mSize` is also 0, and `% 0` should never happen.

Reply:
ack